File: ruby/lib/filestore/warehouse_reader.rb
require_relative './data_file_exception'
require_relative './stock_memento'

module Filestore

	class WarehouseReader

		def initialize(filename)
			@filename = filename
		end

		def read_all
			lines = []
			begin
				File.foreach(@filename) do |line|
					args = line.split(/\s+/)
					lines << StockMemento.new(args[0], args[1], args[2], args[3])
				end
			rescue Exception => e
				raise DataFileException.new(@filename, e)
			end
			lines
		end

	end
end

File: python/filestore/warehouse_reader.py
from stock_memento import StockMemento

class WarehouseReader:

    def __init__(self, filename):
        self.filename = filename

    def read_all(self):
        lines = []
        try:
            fp = open(self.filename)
            for i, line in enumerate(fp):
                args = line.split()
                lines.append(StockMemento(args[0], args[1], args[2], args[3]))
        except IOError as e:
            raise DataFileException(self.filename, e)
        finally:
            fp.close()
        return lines

